### ISSUE TITLE
Fix various symbol names and other minor typos

### ISSUE THREAD

**Issue by smithery1** [virtualagc/virtualagc#1143]
I've written an AGC assembler to learn a little more about the computer. It's based solely on my reading of the original MIT/IL documentation that Virtual AGC has collected, but of course takes yaYUL-formatted source files as input.

For fun, it also outputs end-of-listing information that mimics the GAP output on the original scans. I've spot-checked those listings against Comanche055 and Luminary099, which mostly involves comparing entries between my output and the scans, or the first and last entry on each page for large tables, or the first and last entry on every Nth page for very large tables.

The spot checks have identified a handful of discrepancies in places like location names and COUNT operands. These are mostly minor transcription errors in symbol names, causing them to appear out of order in the EBCDIC-sorted informational tables. I've verified the proper symbol names by the symbol's location in the original tables, and by staring closely at the scans for those cases where the sort problem is a result of a character change to one with a similar glyph (ZEROD to ZEROO or P00DAPAD to POODAPAD for example.) There were also added a couple of unreferenced symbols that were omitted from the transcriptions.

These don't effect the binary output, of course, but they are deviations from the original and do impact auxiliary information like these tables. 

Proposal here is to more comprehensively identify such issues across all code bases, including the comanche67 branch.  The goal is to keep the code bases cross-consistent wrt any changes.

Not sure I can assign or be assigned this Issue, but I'll work on it. Any complications that come up can be discussed here, and I'll post a summary of the findings prior to submitting a PR.

**Comment by rburkey2005** [comment]
@smithery1 By the way, I didn't mention it in our other message stream, but the idea of looking at the EBCDIC sort ordering to find symbol misspellings is terrific, and I feel kind of stupid for not having thought of it, and for not implementing an EBCDIC sort-order option in the yaYUL assembler.  In particular, the problem of resolving 0 vs O in the printouts themselves, given their generally miserable quality, is very tough when all you can do is eyeball it.  They look almost exactly the same. But using the original sorting order and then comparing the newly-assemble symbol tables to the ones in the original printout resolves that question almost completely.  (Slap forehead!)  Well, it's not the first time I felt stupid.  :smile:

I've "assigned" the issue to you, for whatever that's worth.

**Comment by smithery1** [comment]
Thanks - I wish I could say that was my intent in recreating the tables, but it does turn out that they are a gold mine for finding things like this. And generating their content was a much less obscure process than I feared. (They also make heavy use of page numbers, and I was very happy to see those comments in the transcribed source.)

**Comment by rburkey2005** [comment]
Yes, my original intention in writing yaYUL was to allow development of AGC code in a "modern" environment, using modern text editors, so I didn't bother to try and recreate stuff like columnar alignment of the source code, comments without comment markers ('#'), identical formatting of the output assembly listings, and of course EBCDIC.  Actually, when I wrote yaYUL, I didn't even know that EBCDIC had been originally used.  So I think I succeeded in what I was attempting to do, but the "improvements" have definitely turned out to have negative trade-offs.  If you're not aware, the original YUL code is being resurrected, as well as independently being recoded into python, either of which might have allowed us to notice the same things you did -- if we were sufficiently sharp witted!  Those efforts are hampered by my formatting "improvements", since the source code is no longer in a format accepted by the original YUL.

But I digress.  My point, such as it was, is that seemingly inconsequential choices (like character encoding) can have large unintentional consequences later on.  Oh, well!

**Comment by smithery1** [comment]
It makes sense, there's always the question of whether a particular fidelity is worthwhile or will turn out to be more of an annoyance. There are a lot of punch card details that probably would've just obscured the code.

The nice thing about the yaYUL formatting is that it can mostly work either way re: columns. Because of YUL's column-oriented approach, I believe it accepted all sorts of things a standard whitespace-ignoring LR parser would consider goofy, such as embedded spaces in symbols, symbol names that matched opcodes, etc. So something like "TC TC" - is that a "TC" location symbol on an infinite loop, a transfer to a "TC" location, or a IAW for the symbol "TC TC"? Being somewhat paranoid about that, I lexed the yaYUL fields on (somewhat loosely defined) column boundaries, and it works! (Although fortunately I believe the developers used sensible naming everywhere anyway.)

I saw the YUL code scans and planned to go back and look it over in detail when I was more familiar with things. (Trying to learn ARGUS/MITIGUS to try to learn the assembly process to try to learn the AGC language seemed like too much at first. Although I couldn't help but throw in cusses and a version of The Great Big Bad Secret in my simple assembler.) What does the "resurrection" of the original code entail? There are a few small source code quirks that empirically disagree with the YUL documentation, some other things it doesn't completely specify, and cross-documentation inconsistencies, and it'd be interesting to see what the version of YUL being resurrected supported vs. the 1963 documentation vs. other documentation vs. the GAP-accepted code.

**Comment by rburkey2005** [comment]
I don't know if YUL itself accepted symbols with embedded spaces, but I believe that at some point there was a plan for it to do so, and I think that ARGUS does accept such symbols.  As you note, the actual AGC source code does not have this perplexing "feature", which is quite fortunate since I didn't have copies of YUL or any supporting documentation for it what I wrote yaYUL.

By the "resurrection" of YUL, I meant just that the digitized hardcopy of YUL's assembly listing has been converted to machine-readable source code, and that there's an attempt to assemble it and run YUL to actually assemble an AGC program.  Which unfortunately means that it's also necessary to simulate the computer on which YUL originally ran, and at least portions of whatever operating system that computer had.  It's a bit Quixotic, and will certainly be a tour de force if successful, but I'm not personally involved and don't know the current status.  Also, since almost all of the AGC source code we have was written for GAP, I seem to vaguely recall that some mods to the AGC source are needed for YUL, and that not all versions of the AGC source code may be compatible.  I assume it's a very long-range product.  @jimlawton  is the one doing this.

The (separate) attempt to transcode YUL into python is a way of getting the precise functionality of YUL but without simulating its computing platform.  I don't know the full status of that attempt either, but I recall that it's at least partially functional.  @thewonderidiot is the one doing this.

Those are separate github repositories from the virtualagc repo.

**Comment by smithery1** [comment]
Got it, thanks, those sound interesting.

**Comment by thewonderidiot** [comment]
The python port of Yul lives here: https://github.com/thewonderidiot/pyul 
It's for the most part a direct translation of the H-1800 Yul listing to pythonic "pseudocode". I made a few modern simplifications like using a map-like structure instead of linked lists for the symbol table, but for the most part the actual meat of the code doing the assembly should match quite closely.

As of right now, the AGC4 and BLK2 computers are fully functional, meaning that Solarium, Retread, and Aurora can all be assembled using it. I've also translated all three of those to punch card format (in the [decks](https://github.com/thewonderidiot/pyul/tree/master/decks) folder). Due to the differences noted, it takes a `--year` flag to switch on and off different features of Yul to make it mimic how Yul behaved in that particular timeframe. The same flag, or possibly a `--gap` one, will also do the formatting changes between Yul and GAP, once I make it that far through the punchcard endeavor.

Through punchcard creation I've also been painstakingly proofreading everything and making corrections as I go... which I've been _mostly_ mirroring in the virtualagc source locally, where the differences aren't only whitespace. But I've been dragging my feet on committing all of that and pushing it upstream, which I really need to get around to doing. I've also been thinking about making a punchcard-deck-to-yaYUL-format generator script, which would go from these strictly-formatted punchcards back into yaYUL format... allowing all of the whitespace changes to be brought over as well, if we standardized on a way to handle the different types of remarks cards. That would make all of the yaYUL-formatted versions easily diffable as well, which would _super_ nice.

Anyways, I can answer the spaces-in-symbols question: yes, Yul supported this, although it seems to have been avoided wherever possible in the AGC code. I've only ever found one instance of it, in Solarium:
![image](https://user-images.githubusercontent.com/94056/113602726-cf298e80-95f7-11eb-8541-df6d1fc4833e.png)
![image](https://user-images.githubusercontent.com/94056/113602788-e8323f80-95f7-11eb-98ad-32a0a1f5db4c.png)

It doesn't matter in this case, but for awareness the symbol table for Yul is ordered by Honeywell character value, not EBCDIC.

**Comment by jimlawton** [comment]
Long-range it is :-). I moved to a new job three years ago and my hobby-time disappeared. 
Mike's idea is a super one (as usual :-)), I hadn't seen that.

**Comment by smithery1** [comment]
Ah, I figured the EBCDIC encoding was probably a GAP change, I think the YUL documentation contains a character code cross-reference table that has various IBM and Honeywell sets but no EBCDIC.

I see @jimlawton at one point adjusted that "ROOT 2/4" to "ROOT2/4" in the yaYUL source. 😄

I just had time for a short glance, but the pyul project seems very interesting. The Mylar tape output is probably something no one has seen in a long time.

I would like to continue with this Issue and do some table-based reconciliation of symbols and other various non-binary-impacting clerical card data, but I don't want to unduly conflict with your local changes. I think it all should merge smoothly unless one of us was wrong, which would be good to know anyway. It's fine of course if my changes are eventually replaced with data generated from the cards (which does sound nice), but they'd be good to have in the short-term I believe. Let me know if you have any concerns.

**Comment by thewonderidiot** [comment]
> Ah, I figured the EBCDIC encoding was probably a GAP change, I think the YUL documentation contains a character code cross-reference table that has various IBM and Honeywell sets but no EBCDIC.

Yep!

> I just had time for a short glance, but the pyul project seems very interesting. The Mylar tape output is probably something no one has seen in a long time.

I haven't implemented the `RAYTHEON WIRING TAPE` or `RAYTHEON TESTER TAPE` subdirectors for any of the computers yet, although the desire to see/understand those was one of the motivating factors for pyul. :slightly_smiling_face: As of now I've only implemented the basic structure of `MANUFACTURING`, and have added the new subdirector `PUNCH BINARY` (or optionally, `PUNCH BINARY FOR VIRTUALAGC`) since that's more immediately useful.

> I would like to continue with this Issue and do some table-based reconciliation of symbols and other various non-binary-impacting clerical card data, but I don't want to unduly conflict with your local changes.

Please go ahead, by all means! I'll gladly eat any additional work I suffer as a result of not having been proactive about pushing my changes, haha. And more eyes looking for errors is always good! :smile:

**Comment by smithery1** [comment]
> I haven't implemented the `RAYTHEON WIRING TAPE` or `RAYTHEON TESTER TAPE` subdirectors for any of the computers yet, although the desire to see/understand those was one of the motivating factors for pyul.

I'll be following with interest. I can think of all sorts of impractical but fun things to do with the Raytheon tapes once that data is available and understood. Next up, yaLOL!

> Please go ahead, by all means!

Sounds great!

**Comment by smithery1** [comment]
Hello again. My assembler is now successfully processing the code bases on the LM branch of the AGC tree, from Sunburst 37 though Luminary 210, including Zerlina 56, so it seemed like a good time to record what I encountered along the way. There's a fair amount of data here for journaling purposes, feel free to read or ignore the details.

### Approach
Once my assembler was generating the correct binary output for Luminary 99, I visually compared every entry in each end of listing table (except the octal one) to the scans. For each discrepancy, I compared the symbol's occurrences in the scans to the transcription, and corrected the transcription as necessary.

I repeated this process with the other code bases, but less exhaustively - I checked only the first entry in each column on each page of the tables, along with the final entries in the tables. That was enough to spot many name issues that resulting in a misordering, but there may be others still left undetected. (The original YUL output of the Sunburst 37 Cross-Reference tables in the scans is kind of an interesting mess for a couple of reasons and I didn't check those too closely.)

My assembler no doubt has a slightly different set of things it complains about than yaYUL, and it had occasional issues with the code. Most of those were minor problems of field alignment or punctuation. Each was again checked against the scans and transcribed text updated as appropriate.

All corrections were applied consistently across code bases where required.

### Validity of Changes
I am confident that the transcription issues identified were corrected appropriately. Each individual edit was validated manually before being applied. Where scans are available, the edit was compared to the image. Where scans are not available, I examined the code alongside the code from which it was reconstructed (and their scans) to make a determination. Any non-trivial cases are annotated in the sets of changes below.

A script containing the running collection of corrections was used to flag similar corrections across lines, files, and code bases. It did not apply the edits automatically, but provided editor macros to locate and update them one at a time. This allowed for validation of each edit and ensured a consistent update for each change.

### Binary Integrity
Code bases were regression tested for binary correctness by a script that performs a multi-way comparison between the yaYul bin output on an untouched master repo, the yaYul bin output on the modified fork, my assembler output on the modified fork, and the binsource data on the master repo (where available.) All agree across all code bases. (Well, with one exception unrelated to my changes. See Unresolved below.)

### Tables
A few tables are included below that cover each change against each revision. They're probably too big to take in all at once with github's narrow formatting, but they allow anyone interested to pick out all changes for a code base or all code bases affected by a change.

The Basis column contains a key explaining the reason for and validation of the change.
- **A**: Found via an assembler error and validated by visual inspection of scans and/or code
- **CT**: Proper location in the sorted "Count Data" table
- **ST**: Proper location in the sorted "Symbol Table" table
- **UT**: Proper location in the sorted "Unreferenced Symbol" table
- **V**: Found and validated by visual inspection of scans and/or code

Several tables use a key for a change's applicability to a code base.
- **C**: Correct as is. The change in this row was already correctly transcribed for the code base in this column
- **N**: N/A. The change in this row was not applicable to the code base in this column, i.e. the correction was obsoleted by a subsequent change in the code base
- **R**: Revised. The change in this row was applied to the code base in this column

The code base columns are marked with the following symbols as a note of their origin.
- (No symbol): From scans
- **\***: Reconstructed
- **+**: Disassembled / reconstructed

**Comment by smithery1** [comment]
### Unresolved

A few issues that came up that I could not resolve immediately.

- **yaYUL assembly of Sunburst 37**
    I believe yaYUL has trouble with the two unaltered lines below from SECOND_DPS_GUIDANCE, which the virtualagc comments in the code note are syntax modifications from the original source for yaYUL compatibility. I ran a vanilla version of yaYUL on unmodified master Sunburst37 files, and it output a zero word for each of the lines in question. Everything else was correct (mod bugger word for that bank). It could well be something wrong with my setup, but it seemed odd and I'm mentioning it in case it's of interest.
	```
            DEC      1.953125   E-3        B14
            DEC      -3.90625   E-2        B14
	```
	Replacing the lines with the simple octal constants "OCT 40" and "OCT 76577" works, but I left them alone. I'm trying my best to avoid assembly spoilers by looking at the yaYUL code, but when I've assembled everything myself and have some time later I may come back to this if it's still an issue.
	
- **LOWSUPER in Sunburst 120**
    A comment in ERASABLE_ASSIGNMENTS marks this as an addition to the original code to work around an assembler discrepancy. It causes some grief for my assembler and a symbol table difference, and omitting it does not change the yaYUL output in any way (no command-line flags used.) Is it still needed?

- **EBANK= in Sundance revisions**
    There may be a few issues with matching EBank addresses to `EBANK=` settings in Sundance XXX and 306ish, which would require some `EBANK=` directives to correct, understandable given the disassembly origin of the code. I see other such directives in the code, so I didn't look very closely and left these alone for now pending further investigation. I may take a closer look later.

**Comment by smithery1** [comment]
### Symbol Changes

Changes in symbol names.

Symbol|Corrected|Basis|Notes|S37|S120|SXXX+|S306ish+|L69|L97\*|L98\*|L99|L116|L130\*|L131|L163\*|L173\*|L178\*|Z56|L210
|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-
24NDEND|2RNDEND|ST||C|C|C|C|C|C|C|C|C|R|R|C|C|C|C|C
B5TOBB|B5TOB8|V||R|R|R|R|R|R|R|R|R|R|R|R|R|R|R|R
CANTROD|CANTROO|V||N|N|C|C|C|R|R|R|R|R|R|R|R|R|R|R
DELTAOK|DELTATOK|ST||N|N|R|R|R|R|R|R|R|R|R|R|R|R|R|R
DFC&#8209;12|DEC&#8209;12|ST||N|N|R|R|R|R|R|R|R|R|R|R|R|R|R|R
DFC&#8209;6|DEC&#8209;6|ST||N|N|R|R|R|R|R|R|R|R|R|R|R|R|R|R
GTSGO+DN|GTSGO+ON|V||N|N|R|R|R|R|R|R|R|R|R|R|R|R|R|R
ISWCALLL|ISWCALL|V||R|R|R|R|R|R|R|R|R|R|R|R|R|R|R|R
K10RK2|K1ORK2|ST||N|N|R|R|R|R|R|R|R|R|R|R|R|R|R|R
LADFWDV|LATFWDV|ST||N|N|C|C|C|C|C|C|C|R|R|R|R|N|N|N
LUROY|LUROT|CT|COUNT\*&nbsp;operand|N|N|C|C|C|C|C|C|C|R|R|C|C|C|C|C
MAXISHIFT|MAXISHFT|ST||N|N|R|R|R|N|N|N|N|N|N|N|N|N|N|N
MINISHIFT|MINISHFT|ST||N|N|R|R|R|N|N|N|N|N|N|N|N|N|N|N
NEGTORKP|NEGTOTKP|V|See&nbsp;NEGTORKP&nbsp;note&nbsp;below|N|N|N|N|N|N|N|N|N|N|N|R|R|R|C|C
NETAPOSB|NEGAPOSB|ST||N|C|N|N|N|N|N|N|N|N|N|N|N|N|N|N
NETTOTKP|NEGTOTKP|ST|See&nbsp;NEGTORKP&nbsp;note&nbsp;below|N|N|N|N|C|R|R|R|R|R|R|N|N|N|N|N
OCT21103|OCT1103|ST|Value&nbsp;and&nbsp;symbol&nbsp;change&nbsp;to OCT21103&nbsp;in&nbsp;116|N|N|C|C|R|R|R|R|N|N|N|N|N|N|N|N
P00FIZZ|POOFIZZ|ST||N|N|C|C|C|R|R|R|C|R|R|C|C|C|C|C
P00H|POOH|ST||C|C|C|C|C|R|R|R|C|C|C|C|C|C|C|C
P00KLEAN|POOKLEAN|ST||N|N|C|C|C|R|R|R|C|R|R|C|C|C|C|C
R00|ROO|ST||N|N|C|C|C|R|R|R|C|N|N|C|C|C|C|C
R00AD|ROOAD|ST||N|N|C|C|C|R|R|R|C|R|R|C|C|C|C|C
R21LEMB|R21LEM8|ST||N|N|N|N|N|R|R|R|R|R|R|C|C|C|C|C
r29|R29|CT|COUNT\*&nbsp;operand|N|N|C|C|C|R|R|R|C|R|R|C|C|C|C|C
R29REM0J|R29REMOJ|V|Remark&nbsp;OWN&nbsp;JOB|N|N|N|N|R|R|R|R|R|R|R|N|N|N|N|N
ROC|ROO|V||N|N|C|C|C|N|N|N|C|R|R|C|C|C|C|C
SEUDOP00|SEUDOPOO|V||N|N|C|C|C|R|R|R|C|R|R|C|C|C|C|C
SINSLOC|SINESLOC|ST||N|N|R|R|R|R|R|R|R|R|R|R|R|R|R|R
STARTMNAD|STRTMNAD|ST||N|N|N|N|N|N|N|N|N|N|N|N|N|N|R|R
TGDCALC|TGOCALC|ST||N|N|R|R|R|R|R|R|R|R|R|R|R|R|R|R
TICKTIFF|TICKTFF|V|Remark&nbsp;TICK&nbsp;TFF|N|N|R|R|R|R|R|R|R|R|R|R|R|R|R|R
TTPIO|TTPI0|V||N|N|R|R|R|R|R|R|R|R|R|R|R|R|R|R
UPCK|UPOK|ST||R|R|R|R|R|R|R|R|R|R|R|C|C|C|C|C
W.INDI|W.IND1|V|Card&nbsp;is&nbsp;EQUALS&nbsp;W.IND&nbsp;+1|N|N|N|N|R|R|R|R|R|R|R|N|N|N|N|N
ZEROD|ZEROO|ST||N|N|C|C|C|R|R|R|R|R|R|R|R|R|R|R

#### NEGTORKP
The original source code typo `NEGTOTKP` was changed to the correct `NEGTORKP` sometime in Luminary revisions (131, 210]. Its appears in the code bases as follows.
* Luminary 131 and earlier - defined as typo'd `NEGTOTKP` but never referenced
* Zerlina 56 (initially branched from Luminary 145) - has corrected `NEGTORKP`, references it in an `=` and a `CHECK=` statement, also removes the `USED EXCLUSIVELY FOR DOWNLIST` remark
* Luminary revisions [163, 178] (reconstructed) - usage identical to Luminary 131 and earlier
* Luminary 210 - usage identical to Zerlina 56

So when was the typo likely corrected and should Luminary revisions [163, 178] have the typo or the correction? Given its long uncorrected history I believe it would have been noticed and fixed when the code was added that referenced the definition. The apparent contradiction in the list above is that Zerlina, initially branched from Luminary 145, has the references, but later Luminary revisions [163, 178] do not have references. However, this is answered by Luminary Memo #177, which states that Zerlina 56 contains the changes up to Luminary 183.

Therefore it appears that the references and typo correction happened somewhere in Luminary revisions (178, 183]. Given that, I've reverted to the `NEGTOTKP` spelling in Luminary revisions [163, 178].

**Comment by smithery1** [comment]
### Additions and Removals

Various symbols and clerical card additions and removals.

Symbol|Location|Basis|Notes|S37|S120|SXXX+|S306ish+|L69|L97\*|L98\*|L99|L116|L130\*|L131|L163\*|L173\*|L178\*|Z56|L210
|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-
BANK&nbsp;16 EBANK=&nbsp;DT|P&#8209;AXIS_REACTION_CONTROL_SYSTEM_AUTOPILOT&nbsp;P492|A|Lines&nbsp;present&nbsp;in&nbsp;scans,&nbsp;missing&nbsp;from&nbsp;transcription|C|A|N|N|N|N|N|N|N|N|N|N|N|N|N|N
BODYFLAG|ERASABLE_ASSIGNMENTS|ST|Location&nbsp;symbol&nbsp;present&nbsp;in&nbsp;scans|A|N|N|N|N|N|N|N|N|N|N|N|N|N|N|
DEC66|LUNAR_LANDING_GUIDANCE_EQUATIONS|ST|Not&nbsp;present&nbsp;in&nbsp;69,&nbsp;required&nbsp;in&nbsp;code&nbsp;for&nbsp;97+|N|N|R|R|R|N|N|N|N|N|N|N|N|N|N|N
EBANK=&nbsp;DT|Q,R&#8209;AXES_REACTION_CONTROL_SYSTEM_AUTOPILOT&nbsp;P519||Non&#8209;one&#8209;shot&nbsp;line&nbsp;is&nbsp;present&nbsp;in&nbsp;scans,&nbsp;missing&nbsp;from&nbsp;transcription|N|A|N|N|N|N|N|N|N|N|N|N|N|N|N|N
EBANK=&nbsp;JETSTEP|LEM_FLIGHT_CONTROL_SYSTEM_TEST&nbsp;P373|A|Line&nbsp;is&nbsp;present&nbsp;in&nbsp;scans,&nbsp;missing&nbsp;from&nbsp;transcription|A|A|N|N|N|N|N|N|N|N|N|N|N|N|N|N
ITR0|INTERPRETER&nbsp;|ST|Location&nbsp;symbol&nbsp;present&nbsp;in&nbsp;scans|C|C|A|A|A|A|A|A|A|A|A|C|C|C|C|C
R21LEM10|P20&#8209;P25|ST|Not&nbsp;present&nbsp;in&nbsp;69,&nbsp;required&nbsp;in&nbsp;code&nbsp;for&nbsp;97+|N|N|N|N|R|N|N|N|N|N|N|N|N|N|N|N
SIDLOOP|AGC_BLOCK_TWO_SELF&#8209;CHECK&nbsp;P1078|ST|Location&nbsp;symbol&nbsp;not&nbsp;present&nbsp;in&nbsp;scans|C|R|N|N|N|N|N|N|N|N|N|N|N|N|N|N
SPSCODE&nbsp;=&nbsp;BIT9|IMU_MODE_SWITCHING_ROUTINES|ST|Line&nbsp;not&nbsp;present&nbsp;in&nbsp;scans|C|R|N|N|N|N|N|N|N|N|N|N|N|N|N|N
WO1|DOWN&#8209;TELEMETRY_PROGRAM|ST|Not&nbsp;present&nbsp;in&nbsp;69,&nbsp;present&nbsp;in&nbsp;99. 97&nbsp;code&nbsp;similar&nbsp;to&nbsp;69,&nbsp;98&nbsp;code&nbsp;similar&nbsp;to&nbsp;99.|C|C|R|R|R|R|N|N|N|N|N|N|N|N|N|N

**Comment by smithery1** [comment]
### Reference Changes

The four symbols here were defined but unreferenced in Luminary 69; the equivalent BIT# symbol was used in instruction operands instead. By Luminary 99, the BIT# operands had been replaced with the more descriptive names. So it makes sense that they would be unreferenced in the Sundance versions as well. The Luminary 97 and 98 code sections that reference the descriptive names are more similar to 99 than 69, so those were left alone.

Referenced|Corrected|Basis|S37|S120|SXXX+|S306ish+|L69|L97\*|L98\*|L99|L116|L130\*|L131|L163\*|L173\*|L178\*|Z56|L210
|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-
ANTENBIT|BIT12|UT|N|N|R|R|R|N|N|N|N|N|N|N|N|N|N|N
AUTOMBIT|BIT2|UT|N|N|R|R|R|N|N|N|N|N|N|N|N|N|N|N
RCDUFBIT|BIT7|UT|N|N|R|R|R|N|N|N|N|N|N|N|N|N|N|N
REMODBIT|BIT14|UT|N|N|R|R|R|N|N|N|N|N|N|N|N|N|N|N

**Comment by smithery1** [comment]
### Page Numbering Changes

The end of listing tables make heavy use of the page numbers, of course, and I corrected a few places where they were off. This includes reconstructed versions 130, 163, 173, and 178. Hopefully that makes sense to you as well - the tables are interesting for those revisions too. The renumberings were all done by a script and were manually verified against the scans where available.

Action|Notes|S37|S120|SXXX+|S306ish+|L69|L97\*|L98\*|L99|L116|L130\*|L131|L163\*|L173\*|L178\*|Z56|L210
|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-
##PAGE302&nbsp;&#8209;>&nbsp;##&nbsp;Page&nbsp;302|ALARM_AND_ABORT.agc|X|||||||||||||||
#&nbsp;Page&nbsp;447&nbsp;&#8209;>&nbsp;##&nbsp;Page&nbsp;447|IMU_PERFORMANCE_TESTS_3.agc||X||||||||||||||
804&#8209;>797,&nbsp;984&#8209;>976,&nbsp;1215&#8209;>1206, 1216&#8209;>1207,&nbsp;1259&#8209;>1249, 1432&#8209;>1422,&nbsp;1452&#8209;>1453||||||||||X|||||||
UNIX&nbsp;style&nbsp;line&nbsp;endings: INFLIGHT_ALIGNMENT_ROUTINES,&nbsp;POWERED_FLIGHT_SUBROUTINES|Side&nbsp;effect&nbsp;of&nbsp;page&nbsp;renumbering&nbsp;script&nbsp;but&nbsp;now matches&nbsp;other&nbsp;files|||||||||X|||||||
Add&nbsp;empty&nbsp;page&nbsp;for&nbsp;SUBRO&nbsp;output at&nbsp;end&nbsp;of&nbsp;FLAGWORD_ASSIGNMENTS|Now&nbsp;matches&nbsp;Luminary099,&nbsp;but&nbsp;note&nbsp;other&nbsp;code bases&nbsp;such&nbsp;as&nbsp;Luminary069&nbsp;and&nbsp;Luminary116 include&nbsp;the&nbsp;subroutine&nbsp;references||||||||||X|X|||||
Renumber&nbsp;most&nbsp;pages|The&nbsp;pages&nbsp;are&nbsp;out&nbsp;of&nbsp;order,&nbsp;probably&nbsp;due&nbsp;to&nbsp;the reconstructed&nbsp;nature&nbsp;of&nbsp;the&nbsp;code&nbsp;base.&nbsp;A sampling&nbsp;of&nbsp;pages&nbsp;across&nbsp;revisions&nbsp;seems&nbsp;to indicate&nbsp;that&nbsp;the&nbsp;page&nbsp;break&nbsp;points&nbsp;are appropriate,&nbsp;just&nbsp;not&nbsp;correctly&nbsp;numbered.||||||||||||X|X|X||

**Comment by smithery1** [comment]
### Other Changes

Miscellaneous updates that do not fit in the previous categories. Code base columns contain pages that changed, or file lines where page numbers are not available.

Other|Corrected|Basis|Location|Notes|S37|S120|SXXX+|S306ish+|L69|L97\*|L98\*|L99|L116|L130\*|L131|L163\*|L173\*|L178\*|Z56|L210
|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-
Extended&nbsp;numeric&nbsp;constant&nbsp;fields|Fix&nbsp;single&nbsp;\*&nbsp;uses|A|SECOND_DPS_GUIDANCE||P846,&nbsp;P867|||||||||||||||
Misc&nbsp;IAWs|Align&nbsp;to&nbsp;address&nbsp;field|A|INFLIGHT_ALIGNMENT_ROUTINES|To&nbsp;match&nbsp;alignment&nbsp;on&nbsp;scans|Various|Various||||||||||||||
Extended&nbsp;numeric&nbsp;constant&nbsp;fields|Fix&nbsp;single&nbsp;\*&nbsp;uses|A|LUNAR_LANDING_GUIDANCE_EQUATIONS||||L1133,&nbsp;L1139|L1133,&nbsp;L1139||||||||||||
+8|+8D|A|FRESH_START_AND_RESTART|Per&nbsp;YUL&nbsp;docs&nbsp;section&nbsp;13&#8209;114,&nbsp;it&nbsp;would&nbsp;treat&nbsp;non D&nbsp;terminated&nbsp;numbers&nbsp;with&nbsp;non&#8209;octal&nbsp;digits&nbsp;as decimal&nbsp;but&nbsp;would&nbsp;warn|||L126,&nbsp;L127|L135,&nbsp;L136||||||||||||
IAW&nbsp;LOADTIME|Align&nbsp;to&nbsp;address&nbsp;field|A|R31||||L130|L132|P714|||||||||||
TCF&nbsp;BURNBABY|Align&nbsp;to&nbsp;operand&nbsp;field|A|P40&#8209;P47|Also&nbsp;shift&nbsp;surrounding&nbsp;comments&nbsp;to&nbsp;match|||L139|L141|P761|||||||||||
...&nbsp;DDV\*&nbsp;/&nbsp;APSVEX1|...&nbsp;DDV\*&nbsp;/&nbsp;APSVEX1,1|A|SERVICER|See&nbsp;APSVEX&nbsp;note&nbsp;below|||L182|L185||||||||||||
Extended&nbsp;numeric&nbsp;constant&nbsp;fields|Fix&nbsp;single&nbsp;\*&nbsp;uses|A|LUNAR_AND_SOLAR_EPHEMERIDES_SUBROUTINES||||L202|L202||||||||||||
TCF&nbsp;B\*RNB\*B\*|Align&nbsp;to&nbsp;operand&nbsp;field|A|P40&#8209;P47|Also&nbsp;shift&nbsp;surrounding&nbsp;comments&nbsp;to&nbsp;match|||L257|L259|P763|||||||||||
Extended&nbsp;numeric&nbsp;constant&nbsp;fields|Fix&nbsp;single&nbsp;\*&nbsp;uses|A|CONTROLLED_CONSTANTS||||||P54||||||||P52|P52||P53
3|Remove&nbsp;typo&nbsp;line|A|FLAGWORD_ASSIGNMENTS||||||P92|||||||||||
20'35&nbsp;OCT.&nbsp;28,1968|16:27&nbsp;JULY&nbsp;14,1969|V|Main.annotation|||||||||X||||||||
)|Remove&nbsp;typo&nbsp;character|A|KEYRUPT,_UPRUPT|||||||||||P1331|P1331|||||
RCSMONIT&nbsp;EQUALS&nbsp;RCSMON|Add&nbsp;missing&nbsp;line,&nbsp;update&nbsp;references|ST|RCS_FAILURE_MONITOR|||||||||||P193|P193|||||
+8|+8D|A|P&#8209;AXIS_RCS_AUTOPILOT|Is&nbsp;+8D&nbsp;in&nbsp;131&nbsp;and&nbsp;210||||||||||||P1410|P1410|P1410||
RUPTSTOR&nbsp;EQUALS&nbsp;ITEM3|Remove&nbsp;line,&nbsp;not&nbsp;in&nbsp;scan|ST|ERASABLE_ASSIGNMENTS|||||||||||||||||P97

#### APSVEX
I believe this snippet - an indexed instruction without an indexed IAW - is not legal syntax.
```
         VSR4    DDV*
                 APSVEX1
```

The lines above it take care to load the X1 register with a 0 or 1 based on "WHICH VEX SHOULD BE USED", and this uses a "VEX", so indexing on X1 seems appropriate. The binary IAW also has the X1 form, and of course making this change does not change the yaYUL output, so I assume yaYUL is assuming an X1 index when none is provided.
(The dumped ropes (B5) binary word doesn't quite match this word, but it also has the X1 form and I would expect that the reconstructed address locations are not necessarily one-to-one with the differently versioned modules.)

**Comment by rburkey2005** [comment]
> * **yaYUL assembly of Sunburst 37**
>   I believe yaYUL has trouble with the two unaltered lines below from SECOND_DPS_GUIDANCE, which the virtualagc comments in the code note are syntax modifications from the original source for yaYUL compatibility. I ran a vanilla version of yaYUL on unmodified master Sunburst37 files, and it output a zero word for each of the lines in question. Everything else was correct (mod bugger word for that bank). It could well be something wrong with my setup, but it seemed odd and I'm mentioning it in case it's of interest.
>   ```
>            DEC      1.953125   E-3        B14
>            DEC      -3.90625   E-2        B14
>   ```
>   
>   
>   Replacing the lines with the simple octal constants "OCT 40" and "OCT 76577" works, but I left them alone. I'm trying my best to avoid assembly spoilers by looking at the yaYUL code, but when I've assembled everything myself and have some time later I may come back to this if it's still an issue.

What the modern comments added to the source code are trying to say &mdash; apparently inadequately :frowning_face: &mdash; is that yaYUL does not support the syntax used in the original printout, namely
<pre>
OCT     +.001 B14
OCT     -.024 B14
</pre>
because it does not support +/- signs in an `OCT`.  Since this is the only place in the entire AGC codebase where that occurs, it's reasonable not to fix yaYUL and to instead provide alternate pseudo-ops instead.  yaYUL assembles the replacement `DEC`s correctly to (as you say) 00040 and 76577.  I'm at a loss as to why your assembler does not correctly assemble the `DEC`s.  

As for why `DEC`s were chosen as replacements rather than the apparently simpler `OCT 00040` and `OCT 76577` is unclear, since it wasn't documented.  However, it appears as though the original `OCT`s were intended to convey the idea of floats, rather than integers, and therefore using something like `OCT -.024 B14` &rarr; `DEC -3.90625 E-2 B14` rather than `OCT -.024 B14` &rarr; `OCT 76577` would seem to convey that intention better.

**Comment by smithery1** [comment]
Sorry, it's me who wasn't clear with my ambiguous "unaltered" wording. By that I just meant I hadn't changed anything.

I understand the reason for the replacement from the comments, but yaYUL doesn't appear to assemble the *replaced* code properly - it outputs zero words for those two lines. Replacing them with the OCT equivalents results in the proper output. Of course, it may be my setup but I ran it on unaltered master code and was able to assemble all other code bases without error.

**Comment by rburkey2005** [comment]
@smithery1 yaYUL does assemble the replacement `DEC`s in SECOND_DPS_GUIDANCE properly, except on your platform. I checked it before posting my comment. Remind me what your build platform is?  (Not that knowing that actually tells me much that's useful.  :smile:)

> * **LOWSUPER in Sunburst 120**
>   A comment in ERASABLE_ASSIGNMENTS marks this as an addition to the original code to work around an assembler discrepancy. It causes some grief for my assembler and a symbol table difference, and omitting it does not change the yaYUL output in any way (no command-line flags used.) Is it still needed?

Good catch.  It looks like it's no longer needed.

The comments don't document why I thought the definition of `LOWSUPER` was needed in the first place, but I presume it has to do with the fact that for many years yaYUL was unable to guess with 100% accuracy how the original YUL manipulated memory-bank assumptions, and so each version of the AGC source code had to have various extra bank-assumption pseudo-ops sprinkled through it in random places.  Over time, yaYUL was fixed up by several people, and those extra pseudo-ops removed.  The `LOWSUPER` was probably related to those now-removed pseudo-ops in some way, and was neglected because it didn't affect the assembled code other than in the symbol table.

**Comment by rburkey2005** [comment]
Just to be clear about the `DEC`s in SUNBURST 37, here's a cut-and-paste from the relevant lines of the build I did just now:
<pre>
038438,001713: 22,3632           00040                           DEC      1.953125   E-3        B14
038447,001722: 22,3636           76577                           DEC      -3.90625   E-2        B14
</pre>
Moreover, the comparison the build process does of the octals produced by the assembly process vs the octals independently transcribed from the original assembly listing match 100%.

**Comment by smithery1** [comment]
> Just to be clear about the DECs in SUNBURST 37, here's a cut-and-paste from the relevant lines of the build I did just now:

Ah, it figures! Thanks for checking. Here's what I get.

```
038438,001713: 22,3632           00000                           DEC      1.953125   E-3        B14             #
038447,001722: 22,3636           00000                           DEC      -3.90625   E-2        B14             #
```

How odd that it's just those two lines. But I'm running under Cygwin on Windows, so there's probably a whole host of potential toolchain issues on my end. I'll look into it more at some point.

**Comment by rburkey2005** [comment]
> * **EBANK= in Sundance revisions**
>   There may be a few issues with matching EBank addresses to `EBANK=` settings in Sundance XXX and 306ish, which would require some `EBANK=` directives to correct, understandable given the disassembly origin of the code. I see other such directives in the code, so I didn't look very closely and left these alone for now pending further investigation. I may take a closer look later.

I wouldn't worry about this.  Since there's no existing original printout (as far as we know), the only issue is whether the assembled octals match the dump of from the memory cartridges ... which is, of course, is the entire _raison d'etre_ of the attempt to reconstruct that code.

**Comment by rburkey2005** [comment]
> Ah, it figures! Thanks for checking. Here's what I get.
> 
> ```
> 038438,001713: 22,3632           00000                           DEC      1.953125   E-3        B14             #
> 038447,001722: 22,3636           00000                           DEC      -3.90625   E-2        B14             #
> ```
> 
> How odd that it's just those two lines. But I'm running under Cygwin on Windows, so there's probably a whole host of potential toolchain issues on my end. I'll look into it more at some point.

Unfortunate!  I wonder what the difference could be?

I'm sure I have at one time used Cygwin to build Virtual AGC, but it has been many years so I have no particular insight.  (I guess you noticed that build instructions for Cygwin are no longer provided.  Trying to support everything for everyone simply wore me out!)  Possibly there may be a 32-bit vs 64-bit problem.  But why would it affect only those 2 `DEC`s?  There's absolutely nothing out of the ordinary about them, as far as I can see, other than that a ##-comment line precedes them.

Regarding your actual code changes, give me some time to work through them.

**Comment by rburkey2005** [comment]
I wonder if there's some problem with *nix vs Windows line endings ('\n' vs \r\n') in Cygwin, and if the ##-comment somehow interacts badly with that?  The AGC source code uses all *nix line endings.   Sounds like a foolish guess.

**Comment by smithery1** [comment]
> I wouldn't worry about this. Since there's no existing original printout (as far as we know), the only issue is whether the assembled octals match the dump of from the memory cartridges ... which is, of course, is the entire raison d'etre of the attempt to reconstruct that code.

It's a fair point, perhaps it's not worth it. The reason I was thinking of looking is because my assembler validates addressing against the EBANK= setting and warns when they don't match, which is my understanding of what YUL and GAP might have done. This is the only code base for which it complains about EBANK and as you say, it doesn't affect the central issue, but I thought it could be a few extra lines of verisimilitude that would match what the original developers would have had to do.

**Comment by rburkey2005** [comment]
Yes, we certainly want it to be as accurate as we can get it, modulo the syntactical and column-alignment differences between yaYUL and YUL.

Perhaps a more-satisfactory statement would have been that SundanceXXX and Sundance306ish are just the first two steps of a 3-step reconstruction process that _may someday_ result in Sundance306, with luck.  Cleanups of things like `EBANK=` pseudo-ops may be more appropriate after Sundance306 has been completed rather than prior to that point.  @thewonderidiot would probably have more insight into that than I, since he has been working on that reconstruction, on and off.

**Comment by smithery1** [comment]
> I guess you noticed that build instructions for Cygwin are no longer provided.
Yes, understandable! I compiled just yaYUL and oct2bin locally.

> I wonder if there's some problem with *nix vs Windows line endings ('\n' vs \r\n') in Cygwin, and if the ##-comment somehow interacts badly with that? The AGC source code uses all *nix line endings.
Possible. There were two files I think that had Windows line endings that I "corrected" to *nix (see the Page Renumbering table), but nothing had trouble with them before or after. Maybe there are other stray odd line endings around.

> Possibly there may be a 32-bit vs 64-bit problem. But why would it affect only those 2 DECs? There's absolutely nothing out of the ordinary about them, as far as I can see, other than that a ##-comment line precedes them.
Agreed, that's why I found it odd. This is my best guess, perhaps something in my toolchain is 32 bit, though these numbers don't require nearly the precision of many others.

I don't mean to throw some odd incompatibility like this out there without further assistance or information, but I really am enjoying working though assembly issues strictly from the original documentation and am convinced I'll see something great that I don't know yet if I look at yaYUL too closely. However, I can see some light at the end of tunnel, particularly after getting though Sunburst, and I will figure this out soon and post results here.

**Comment by rburkey2005** [comment]
I wasn't suggesting you look in yaYUL.  Even I don't want to do that!  I was really hinting, obliquely, that you might remove the ##-comment lines and see if the problem went away.   But if it's not a big deal, that's fine too.  And probably pointless.

**Comment by smithery1** [comment]
> I wasn't suggesting you look in yaYUL. Even I don't want to do that! I was really hinting, obliquely, that you might remove the ##-comment lines and see if the problem went away. But if it's not a big deal, that's fine too. And probably pointless.

😄 It's mostly curiosity at this point. I verified 64-bit compilation of yaYUL, and I'm continuing to poke around a little outside the source. Thanks for the suggestions.

**Comment by smithery1** [comment]
> after I've finished incorporating the less-challenging problems that have been identified.

I'm not sure if you're making these changes? If so I should have been clearer - this was an FYI and more detailed record about what the proposed changes are - I've made all of them in a branch that I can submit as a PR whenever it's a good time. I thought I'd look at Aurora and the CM code first, but the branch has every change here. If you'd like to play with them I can submit the PR as it stands.

**Comment by rburkey2005** [comment]
I'm afraid that went over my head, or else I forgot it.  When I started reviewing this morning and found no patch, I began making the changes myself.  I saw your repeated comments that you had tested the changes, but I supposed you were taking to heart my statement (about your original, deleted patch) that having a list of the problems was almost worth more than having the patch itself, and didn't intend to post a patch.

So ... if you're going to do a patch later, I'll just discard all of my local changes and eventually review the patch instead.  I only put in a couple of hours worth of effort, so it's not any great loss. 

Just post another message confirming this, before I scrub the mods.

My only comments so far, other than the one about `NEGTORKP`, are:

* The change from `OCT21103` to `OCT1103` was made at Luminary 116, in order to conform to an actual change in the constant.
* Obviously, feel free to continue explaining what you found in as much detail as you like.  It's the kind of thing I would do.  :smile:  Ultimately, though, what really matters is whether a given symbol matches the original printout or not &mdash; where we have the printout.  So I'd be happy even with just a list of the symbols, and _completely_ happy (in the sense of minimizing the checking I'll eventually have to do before accepting a patch) with a paper trail telling me what versions of the scanned code that each of the symbols had been visually checked against.

**Comment by smithery1** [comment]
> Just post another message confirming this, before I scrub the mods.

Yes, I'll post a patch. I didn't want it to cause confusion around it being ready to merge, as I may have additional changes as I go though the other code. But if you don't mind the PR sitting around for a little while, I can attach it now as a work in progress and update it as I make more. Or point you to my forked branch, or get this ready to merge and then do the other code bases, or just leave things as they are, etc.

> Ultimately, though, what really matters is whether a given symbol matches the original printout or not — where we have the printout. So I'd be happy even with just a list of the symbols, and completely happy (in the sense of minimizing the checking I'll eventually have to do before accepting a patch) with a paper trail telling me what versions of the scanned code that each of the symbols had been visually checked against.

I was pretty verbose with the preamble to the tables but to summarize, I validated *every* edit (modification to a line in a file) for *every* change (line in an above table) against *every* scan. No change was made that did not visually match the scan for that version. Even if the change was originally spotted in other some version A and the transcriptions for version B seemed to have the same error, it was still visually confirmed against the scans for version B. If a change was spotted in version A and might reasonably appear in version B but did not,, the scans for version B were checked to confirm its absence. The tables are intended to be that paper trail, the notations in the cells indicate the results of a visual inspection. The "Basis" column indicates how the issue was originally identified.

For reproductions without scans, I examined the code and the surrounding scans to make a conservative determination about whether the change should apply to the reproduction or not, and added an annotation in the tables where that was non-trivial.

Of course, I would encourage you to double-check as much as you'd like and I welcome the questions. If there's something more you think would increase the comfort level around this at any point, let me know.

I'll reply with more detail on the `NEGTOTKP` and `OCT21103` changes separately.

**Comment by rburkey2005** [comment]
No, thanks, I don't need a "work in progress".  It would just cause me to end up duplicating any double-checking when I forgot what I had looked at.

I am, however, in the process at this moment of adding an EBCDIC symbol-table sorting option to yaYUL, and possibly a Honeywell one if it turns out to be needed.  Not that you need it yourself, of course, but it seems useful for the rest of us.

**Comment by smithery1** [comment]
> Perhaps I'm confused (very possible!), but I don't find any place where NEGTORKP was transcribed incorrectly from Luminary 69 through Luminary 210; and reconstructions inherited that same usage. NEGTORKP is (correctly) used as the symbol name everywhere. Some of the original program comments incorrectly reference "NEGTOTKP" instead, and that seems to have been transcribed (correctly, according to my philosophy) as "NEGTOTKP". So the transcribed program comments are inconsistent with the transcribed active code, but agree with the original hardcopy.

I believe the Luminary scans have the definition as `NEGTOTKP` from 69 through 210, where it changes to `NEGTORKP`. The transcriptions, however, have an error in 97 through 131 inclusive in which they use `NETTOTKP` instead of `NEGTOTKP`.

So two overlapping typos for `NEGTORKP`, one original and one transcription.
* `NEGTOTKP`: "TOT" instead of "TOR": Used in the scans from [69, 210)
* `NETTOTKP`: "NET" instead of "NEG": Used in the transcriptions from [97, 210)

Luminary069: Scans: `NEGTOTKP`, Transcribed: `NEGTOTKP` (correct)
![image](https://user-images.githubusercontent.com/46849888/115420136-e2785480-a1c8-11eb-8471-6e3cd4f78cab.png)

Luminary099: Scans: `NEGTOTKP`, Transcribed: `NETTOTKP`
![image](https://user-images.githubusercontent.com/46849888/115418656-a8f31980-a1c7-11eb-8c0b-3e828b9dbf05.png)

Luminary116: Scans: `NEGTOTKP`, Transcribed: `NETTOTKP`
![image](https://user-images.githubusercontent.com/46849888/115418711-b27c8180-a1c7-11eb-8c90-d992182e81e7.png)

Luminary131: Scans: `NEGTOTKP`, Transcribed: `NETTOTKP`
![image](https://user-images.githubusercontent.com/46849888/115418762-be684380-a1c7-11eb-92e0-d075230564ff.png)

Zerlina56: Scans: `NEGTORKP`, Transcribed: `NEGTORKP` (correct)
![image](https://user-images.githubusercontent.com/46849888/115420207-f4f28e00-a1c8-11eb-8694-002f6fa609d6.png)

Luminary210: Scans: `NEGTORKP`, Transcribed: `NEGTORKP` (correct)
![image](https://user-images.githubusercontent.com/46849888/115418882-dd66d580-a1c7-11eb-82ac-8fd40ba2ca4b.png)

I corrected the transcription error in [97, 131]  and reverted the spelling in the reproduced [163, 178]  to  `NEGTOTKP` to match the scans speculatively for the reasons given in the note.

**Comment by smithery1** [comment]
> The change from OCT21103 to OCT1103 was made at Luminary 116, in order to conform to an actual change in the constant.

I tried to express that in my note as well. However, some transcriptions of earlier code bases do not match their scans for this symbol. Here is the Luminary069 scan and transcription for example. 

![image](https://user-images.githubusercontent.com/46849888/115422395-dc837300-a1ca-11eb-8a7a-7c4ceac6cc43.png)

**Comment by smithery1** [comment]
> I am, however, in the process at this moment of adding an EBCDIC symbol-table sorting option to yaYUL, and possibly a Honeywell one if it turns out to be needed. Not that you need it yourself, of course, but it seems useful for the rest of us.

I'm sure you know or will notice, but GAP sorts by EBCDIC and YUL sorts by Honeywell.

**Comment by rburkey2005** [comment]
@smithery1 Also, in Luminary 69 I notice a spurious label `-1` in the transcription which does not appear in the original printout's symbol table.  (Possibly your assembler might not have this.)  It comes from the following lines in the P-AXIS_RCS_AUTOPILOT.agc:
<pre>
...
SELECTYZ           TS       NUMBERT
                   TCF      SELECTP    +1
-1                 TCF      ABORTYZ    +2
JETSALL            OCT      00252
...
</pre>
While that is indeed what the original printout looked like, the column alignment of "-1" is wrong, and it needs to be moved rightward to column 2.  (It's not actually a label, but rather a visual aid to the programmer to help him locate `JETSALL -1`.  Normally these are aligned at column 2+, but the alignment at column 1 fools yaYUL into believing it's a label.)

So, if you could move -1 a bit to the right, I'd appreciate it.  Other than Luminary069, I see the same problem in LUM69R2, Sundance306ish, and SundanceXXX, but not in other versions of AGC.

**Comment by rburkey2005** [comment]
By the way, I'm not trying to systematically find errors in parallel with what you're doing.  I just accidentally noticed these in trying to verify that my new symbol-table sorting is working properly.

But since some of these problems might not show up with your assembler, because weaknesses in yaYUL are being exposed rather than typos in AGC source code, it seems to me that I _will_ eventually have to at least partially repeat what you're doing.  But presumably at that point there will be many less problems to find.  :smile:

**Comment by rburkey2005** [comment]
@smithery1 And another one:  Colossus249, TAGS_FOR_RELATIVE_SETLOC_AND_BLANK_BLANK_CARDS.agc, p. 28.
<pre>
...
R30LOC             EQUALS
P11FOUR            EQUALS
P12A               EQUALS
NORMLIZ            EQUALS
ASENT7             EQUALS
RODTRAP            EQUALS
...
</pre>
The final 4 (`P12A`, `NORMLIZ`, `ASENT7`, `RODTRAP`) are not supposed to be there at all.  (Apparently, I adapted this file from a similar Luminary 131 file, and simply never noticed them.  Fortunately, they did not make their way into any other version of Colossus.)

If you'd remove those 4 lines, it would be splendid.

**Comment by smithery1** [comment]
> I'm reluctant to change the AGC source while you're working with it, so perhaps you'd comment it out along with your other changes?

Of course. I actually got Aurora assembling correctly the other day, so this one is already on my list (along with ENDSPF, which should not be in the transcripts.)

> So, if you could move -1 a bit to the right, I'd appreciate it. Other than Luminary069, I see the same problem in LUM69R2, Sundance306ish, and SundanceXXX, but not in other versions of AGC.

Of course again. My assembler (with its unfair advantage of having been built off the documentation) cheerfully ignores signed numeric location symbols so I may not pick up on any additional ones lurking in the code, but I'll do the ones you pointed out.  

> And another one: Colossus249, TAGS_FOR_RELATIVE_SETLOC_AND_BLANK_BLANK_CARDS.agc, p. 28.

Sure. I think the CM code bases are next for me anyway after RETREAD. I can assemble RETREAD properly (no checksums!), but have yet to verify the symbols there due to yet another YUL end-of-listing variant.

> But since some of these problems might not show up with your assembler, because weaknesses in yaYUL are being exposed rather than typos in AGC source code, it seems to me that I will eventually have to at least partially repeat what you're doing.

One thing that may help is that I can provide you with my generated symbol tables in plain text format when this is all done, and you can diff the yaYUL ones against them. I can't guarantee than mine are completely correct vs the scans, but they match at least at column boundaries on each page. That'll help flag any yaYUL differences, and perhaps other issues.

**Comment by rburkey2005** [comment]
> One thing that may help is that I can provide you with my generated symbol tables in plain text format when this is all done, and you can diff the yaYUL ones against them. I can't guarantee that mine are completely correct vs the scans, but they match at least at column boundaries on each page. That'll help flag any yaYUL differences, and perhaps other issues.

Yes, that probably would be a great shortcut, thanks.  I've just now posted the mods in yaYUL.  Modulo the various symbol errors, I think it works pretty well across the entire spectrum of AGC versions ... with the single exception of RETREAD.  RETREAD seems to have a variable number of symbol-table lines per page (as opposed to the constant 45 lines for the GAP output and 43 lines for the other YUL output), and hence I couldn't get the pagination correct for it.

**Comment by thewonderidiot** [comment]
The rules for symbol table printing around the time of Retread are a bit arcane and took me a bit to work out. It's captured in this logic block in pyul:
```
        if self._old_line.spacing != Bit.BIT1:
            if self._old_line[0] != self._line[0]:
                self._old_line.spacing = 3
                self._sym_liner += ONE_THIRD * 6
            elif (ALPHABET.index(self._old_line[1]) // 16) != (ALPHABET.index(self._line[1]) // 16):
                self._old_line.spacing = 2
                self._sym_liner += ONE_THIRD * 3
```

....which is not clear at all, since I hammered it into shape from the later Yul code, but essentially what it's saying is:
* Every page has 50 lines, blank or otherwise, including the two header lines.
* One symbol is printed per line
* If the first character of the current symbol is different from the last, first print two blank lines
* Otherwise, if the second character of the current symbol is in a different "group" than the last, first print one blank line. A "group" in this case is a set of 16 characters in the H-1800 character set; i.e. characters with values 0-15 form group 0, 16-31 form group 2, etc.
* If any of the blank lines emitted by the previous two cases cause the number of lines on the current page to exceed 50, discard any extra blank lines and emit a page break, then print the current symbol on the third line of the next page, following the two header lines.

**Comment by rburkey2005** [comment]
@thewonderidiot Thanks!  I had inferred a little bit of that from the printout, but certainly would never have come up with the whole thing.

On the other hand ... given that it's only RETREAD at stake, and only its pagination, I think it convinces me to call it quits at this point.  [In the words of John Cleese's Sheriff Langston character in Silverado, "Today my jurisdiction ends here."](https://youtu.be/3SN3lRUoRXA) :smile:

**Comment by smithery1** [comment]
Ah, I had come up with very similar rules for my assembler and can confirm that all the page breaks match the scans for Retread 44. I even used the word "group". I think the only difference was I had 50 lines per page exclusive of the listing header and blank line following it. The rule for initial blank lines on a page isn't triggered for Retread 44, but I believe other tables in other YUL versions seem to use a similar rule and I included that too.

**Comment by rburkey2005** [comment]
If that's true, the simple rules I had inferred from looking at the printouts are incomplete, and I may have to consider a more-complex implementation.  But because I am lazy, I won't really be convinced until the known symbol typos are fixed and I can see for myself that my rules are inadequate.

The rules I had observed exclusive of RETREAD and TRIVIUM were:

- Symbol table has 3 columns.
- 43 rows (of symbols) in YUL, 45 in GAP, but less on the final page of the symbol table.
- On a change of the first character, a separator is printed instead of a symbol.
- If the separator would be the very first thing on a page, it is omitted and the next symbol is printed instead.

**Comment by smithery1** [comment]
The "second character group of 16" rules fortunately only seems to apply to Retread (I haven't looked at Trivium yet.)

My rules for the other tables agree closely with yours, except that I believe an initial separator on a page is allowed.

One other note that is significant for H800 sorting is that all symbols should be assumed 8 characters in length, right padded with the H800 space character as necessary. Since certain printing characters precede the space, this produces an ordering that matches the scans vs. always putting the shorter symbol first.

I applied it to Retread too - if the symbol has length one, use the space code for the "grouping". And also for EBCDIC, although with no printing characters less than space I don't think it makes a practical difference.

**Comment by rburkey2005** [comment]
Thanks, I didn't know about the Luminary099 p1563 example of an initial separator.  I'll make the slight fix to yaYUL for GAP.

And you needn't bother with TRIVIUM.  It has only 5-6 symbols anyway.  But they're all printed in one column rather than being split across 3.  :smile:

**Comment by rburkey2005** [comment]
In an absurd coincidence, I just now pulled up the scanned images from Luminary 99 to double-check.  The viewer opens up on the last page I had looked at, and it was precisely p. 1563.  So the example was right in front of my eyes, but I was too dumb to notice it.  :smile:

**Comment by smithery1** [comment]
(FYI, I can assemble all code bases successfully, so I took at look at the odd problem on my Cygwin platform with yaYUL and the replacement constants in Sunburst37. As a reminder, it assembled the following lines, which were specially crafted to replace an original syntax unsupported by yaYUL, to zero words.

```
DEC             1.953125        E-3 B14
DEC             -3.90625        E-2 B14
```

In brief, Parse2DEC.c calls pow(3) as part of calculating the decimal exponent component of an address field. Cygwin uses a libc implementation called "newlib" instead of glibc, since it seems glibc cannot be ported to Windows easily. The newlib pow code looks substantially less involved than the glibc equivalent and perhaps not surprisingly therefore seems to be less accurate in at least some cases. In particular `pow(10.0, n)` for n=-2 or -3 returns a value that is not *quite* the double representation of 0.01 or 0.001.

That eventually leads to the Parse2DEC comparison ` fmod(x, 1.0) == 0.0` being not *quite* true, and a later failure and an error return value that I would guess is ignored (totally fair given the comment "We don't do a heckuva lot of error-checking in this version") and a zero is output.

I would guess it doesn't happen elsewhere because there just aren't a lot of whole numbers defined in this way. Since you don't support Cygwin, I won't look any further at it, but just wanted to wrap up this minor thread.)

**Comment by rburkey2005** [comment]
Irksome.  However, it looks like all invocations of `pow()` (2 in yaYUL, 1 in yaLEMAP, and 2 in checkdec) all use small-integer exponents (some positive, some negative).  So a correctly-working replacement for `pow()`, to be used only in Cygwin, should be almost-trivial to devise using only regular multiplication and division.  Or perhaps a better approach would be to send a bug-report back upstream to the newlib developers.

**Comment by smithery1** [comment]
Thanks. I did indeed make a change like you suggest locally (it seems enough to always give `pow()` a positive exponent and invert the result if necessary) and yaYUL then works fine across all code bases. I probably will see what the newlib developers think.

**Comment by rburkey2005** [comment]
I've just now pushed my own version of the Cygwin-only workaround for yaYUL, yaLEMAP, and checkdec into the repository.  I was able to test the workaround code, even though I'm not running Cygwin myself.  The only (very slight) uncertainty I couldn't test myself is that __CYGWIN__ is the appropriate predefined macro to use for the conditional compilation.

**Comment by smithery1** [comment]
Thanks for doing this. I am using the "mingw-w64" compiler to build a Windows native version of yaYUL and after some more looking it seems I unfairly blamed newlib. Instead it's something in mingw, or the Windows implementation of pow() in "msvcrt.dll", or somewhere in between. I get the same incorrect results compiling outside of Cygwin using a Windows native mingw-w64 compiler, and correct results when compiling inside of Cygwin using a Cygwin specific gcc compiler. Unfortunately I don't have another Windows native compiler with which to narrow it down further.

If you change the check  in yaYUL.h to the following it all works well. You can also change the comments to note that it's a workaround for the pow() function supplied by mingw-w64.

`
#if defined(__MINGW32__) || defined(__MINGW64__)
`

**Comment by rburkey2005** [comment]
Did you mean `#if defined(__CYGWIN__) || defined(__MINGW64__)`?

**Comment by smithery1** [comment]
No, I meant to replace `#if defined(__CYGWIN__)` with the line above.  Using `__CYGWIN__` would have been appropriate for my original description of the problem, but it isn't defined when using mingw-w64, and the problem exists outside of Cygwin as well. The two `MINGW` defines are for the two compiler target architectures, but it looks like the 64 bit one defines both so you could just replace `__CYGWIN__` with `__MINGW32__` instead.

**Comment by smithery1** [comment]
(Just to be clear, it also works correctly without the `__CYGWIN__` check in Cygwin with the Cygwin gcc compiler, meaning only the `__MINGW32__` check is necessary.)

**Comment by rburkey2005** [comment]
Okay.  But I'm posting it with all 3 macros, even though `__CYGWIN__` is redundant to `__MINGW32__`, since it might be helpful to someone looking at it who would wonder if it still needed to be fixed for Cygwin.  I know from personal experience that the relationship between Cygwin and MINGW isn't always obvious to someone who doesn't work with it a lot.

**Comment by smithery1** [comment]
I agree that it isn't obvious! Just to note, this fix is orthogonal with respect to Cygwin. The `__CYGWIN__` and `__MINGW*__` macros don't overlap.

* gcc under Cygwin defines `__CYGWIN__`, but pow() works fine there without the fix.
* mingw-w64 under Cygwin does not define `__CYGWIN__`, but pow() does not work correctly there.
* mingw-w64 under Windows native does not define `__CYGWIN__`, but pow() does not work correctly there.

So checking for it is not harmful - just replacing the library pow() with your implementation - but not required either since the Cygwin/gcc provided pow() doesn't have any problems.

**Comment by 7ofNine** [comment]
I just checked what the results of pow is for the above two cases using a plain windows/C++ compiler (VS2019 16.10) and haven't seen any problem there. I haven't checked on the binary level yet (a bit tricky) but printed it with a high width and they show the same good result as far as the precision can go with a double (have to try float, too). One general problem is to rely on any comparison for floating point types. This is in general not portable. It works on one machine/implementation but not on the next one. And especially pow is a function that, depending on the details of there implementation, uses the logarithm and exponential which are of course transcendental functions and they have there own numerical issues. Especially using float quickly introduces rounding errors which leads to comparison failures.
Floating point values should never be tested without error limits.

**Comment by rburkey2005** [comment]
A fix is needed rather than a workaround.  Replacing `pow()` is completely inappropriate.  `pow()` is fine.  And the only potential problem is in yaYUL (rather than yaLEMAP and checkdec).  I should have given this more attention.

What the `fmod()` function calls are trying to do is to distinguish between which of two distinct encodings to octal is supposed to be used: an encoding appropriate to integers or else one appropriate to floats.

I'm rolling back all of the `pow()` changes, and looking for a different way of performing the test `fmod()` is now performing.

**Comment by smithery1** [comment]
Sounds appropriate, I was trying not to bug people for an impactful change for something that seemed to be working fine for everyone but me and then in one small case. I didn't even need the `#ifdef`, I could work around it locally.

@7ofNine, for curiosity, can you try this test program and send the output?

```c
#include <math.h>
#include <stdio.h>

int main() {
    int n = -3;
    double val1 = 0.001;
    double val2 = pow(10, n);

    printf("%s\n", val1 == val2 ? "worked" : "failed");
    printf("16384*0.001=%.16f, 16384*pow(10,-3)=%.16f\n", 16384.0 * val1, 16384.0 * val2);
}
```

I get this compiled with gcc under Cygwin:
```
worked
16384*0.001=16.3840000000000003, 16384*pow(10,-3)=16.3840000000000003
```

I get this compiled with mingw-w64 (10.2.0) under Cygwin:
```
failed
16384*0.001=16.3840000000000003, 16384*pow(10,-3)=16.3840000000000039
```

I get this compiled with mingw-w64 (8.1.0) under native Windows:
```
failed
16384*0.001=16.3840000000000000, 16384*pow(10,-3)=16.3840000000000040
```

**Comment by rburkey2005** [comment]
@smithery1  I've just posted what's hopefully a fix.  All of the previous changes are rolled back.  Instead, the `fmod` invocations in yaYUL now have a 1.0E-10 fuzz.  It seems to work properly on both 64-bit and 32-bit Linux.  Give it a try.  Unfortunately, it's a bit of trial-and-error, because I no longer really understand the original test and encodings, since it's been 17-18 years since that code was created.  You may actually understand the issue better than I do, since you've presumably coded the `DEC`/`2DEC` pseudo-op  a _lot_ more recently than I have.  Anyway, I'll keep my fingers crossed.

**Comment by smithery1** [comment]
Works for mingw-w64 (32 and 64 bit) and gcc (64 bit) on Cygwin/Windows! The rule for DEC and 2DEC is that if the address field value must be a whole number if after applying the scaling and factor it is greater than or equal to one. Otherwise it is truncated and a cuss is issued. So one has to check the calculated value for integer-ness in some way, and your epsilon approach here seems pretty standard.

**Comment by 7ofNine** [comment]
@smithery1 
just ran it. This is with the latest SDK under Windows 10 etc.
The result is

worked
16384*0.001=16.3840000000000003, 16384*pow(10,-3)=16.3840000000000003

even if the last digit(the 3) is illusionary.  The IEEE standard says that the decimal precision is 15 digits.
This is for compilation in debug mode. With full optimization in release mode I get the identical result.
My guess is that it is the pow function. Some smart implementations use simple multiplications especially for small integers. This is much faster than the usual blind use of logarithm and exponentiation even if the logarithm to base 2 is surprisingly easy and fast to calculate. And there a lot of rounding errors can creep in. The IEEE standard does not specify a precision for such functions and sometimes even worse order correctness.

**Comment by smithery1** [comment]
Thanks @7ofNine. I guess the mystery lies somewhere in the mingw-w64 compiler then. Agreed that pow (and standard math) implementations can be complex and the safest approach is not to expect an exact result.

**Comment by smithery1** [comment]
Ok, I think the changes for this issue are largely complete. To recap, I compared at least the first entry in every column on every page of every end-of-listing table with the scans, and most changes came from there. Others came from various assembler cusses.
* Every change for a code base with scans matches the scans
* Every change for a code base without scans matches the scans from which the code originated, or as noted in a small number of cases where the change was to derived code

The comparison was not exhaustive, in particular I sampled most of the tables as described and did not validate every entry. But every change was applied across all applicable code bases.

I'll add some more tables here summarizing the CM changes, and apply some minor updates to the LM ones above, then issue the pull request.

**Comment by smithery1** [comment]
### CM Symbol Changes

Symbol|Corrected|Basis|Notes|S55|C237|C249|C044\*|C045\*|M45R2\*|C051\*|C055|A071\*|A072
|-|-|-|-|-|-|-|-|-|-|-|-|-|-
$/CRS61|$$/CRS61|CT|P20&#8209;P25|N|R|R|R|R|R|R|R|C|C
$$/P03|$$/PO3|CT|Likely&nbsp;a&nbsp;typo&nbsp;in&nbsp;the&nbsp;original&nbsp;source|N|R|R|R|R|R|R|R|N|N
$$/T4RUPT|$$/T4RPT|CT|T4RUPT_PROGRAM|N|R|R|R|R|R|R|R|C|C
$$/INTER|03/INTER&nbsp;and&nbsp;00/INTER|CT|INTERPRETER|N|N|N|R|R|R|R|R|N|C
20/SXTMK|10/SXTMK|CT|SXTMARK|N|C|R|R|R|R|R|R|N|N
BLKNR1|BLNKR1|ST||N|N|N|N|N|N|N|N|C|C
CANTR00|CANTROO|V||N|C|R|R|R|R|R|R|C|C
CGF|OGF|ST||N|R|R|R|R|R|R|R|N|N
CKPHI|OKPHI|ST||N|N|R|C|C|C|C|C|C|C
CKTHETA|OKTHETA|ST||N|N|R|C|C|C|C|C|C|C
CKU12|OKU12|ST||N|N|R|C|C|C|C|C|C|C
CKU21|OKU21|ST||N|N|R|C|C|C|C|C|C|C
CKU31|OKU31|ST||N|N|R|C|C|C|C|C|C|C
CMDARMBIT|CMARMBIT|A||N|N|R|R|R|R|R|R|C|C
DFC&#8209;12|DEC&#8209;12|ST||N|C|R|C|C|C|C|C|C|C
DFC&#8209;6|DEC&#8209;6|ST||N|C|R|C|C|C|C|C|C|C
ENDOCM|ENDDCM|ST||N|C|R|R|R|R|R|R|R|R
GONBYBIT|G0NBYBIT|ST||N|R|R|R|R|R|R|R|R|R
ISITP00|ISITPOO|ST||N|C|R|R|R|R|R|R|C|C
MPAC|KMPAC|A|EBANK=&nbsp;in&nbsp;MYSUBS|N|R|R|C|C|C|C|C|C|C
NORMSBIT|NDRMSBIT|ST||N|R|R|R|R|R|R|R|C|C
NOUTINPT|NDUTINPT|ST||N|N|N|N|N|N|N|N|C|C
OCT1DNV|OCTIDNV|ST||R|N|N|N|N|N|N|N|N|N
P00DAPAD|POODAPAD|ST||N|C|R|R|R|R|R|R|C|C
P00FIZZ|POOFIZZ|ST||N|C|R|R|R|R|R|R|C|C
P00H|POOH|ST||N|C|R|C|C|C|C|C|C|C
P00KLEAN|POOKLEAN|ST||N|C|R|R|R|R|R|R|C|C
R00AD|ROOAD|||N|C|R|R|R|R|R|R|C|C
R00TOP00|ROOTTOPOO|||N|C|R|R|R|R|R|R|C|C
ROC|ROO|ST||N|C|R|R|R|R|R|R|C|C
ROOT2/4|ROOT&nbsp;2/4|ST|Left&nbsp;alone&nbsp;for&nbsp;yaYUL&nbsp;compatibility|U|N|N|N|N|N|N|N|C|C
SEUDOP00|SEUDOPOO|V||N|C|R|R|R|R|R|R|C|C
SINSLOC|SINESLOC|ST||N|R|R|C|C|C|C|C|C|C
SP3SPT1|2P3SPT1|ST||N|N|N|N|N|N|N|N|C|C
UPCK|UPOK|ST||C|R|R|C|C|C|C|C|C|C
ZEROD|ZEROO|ST||N|R|R|R|R|R|R|R|R|R
ZEROERROR|ZEROEROR|A||N|N|R|R|R|R|R|R|C|C

**Comment by smithery1** [comment]
### CM Additions and Removals

Symbol|Location|Basis|Notes|S55|C237|C249|C044\*|C045\*|M45R2\*|C051\*|C055|A071\*|A072
|-|-|-|-|-|-|-|-|-|-|-|-|-|-
ENDVBFAN|PINBALL_GAME__BUTTONS_AND_LIGHTS|ST|Symbol&nbsp;not&nbsp;present&nbsp;in&nbsp;scans|R|N|N|N|N|N|N|N|N|N
STORANG|R60,R62|ST|Symbol&nbsp;not&nbsp;present&nbsp;in&nbsp;scans|N|R|R|C|C|C|C|C|C|C
P12A|TAGS_FOR_RELATIVE_SETLOC_AND_BLANK_BANK_CARDS|ST|Line&nbsp;not&nbsp;present&nbsp;in&nbsp;scans|N|C|R|C|C|C|C|C|C|C
NORMLIZ|TAGS_FOR_RELATIVE_SETLOC_AND_BLANK_BANK_CARDS|ST|Line&nbsp;not&nbsp;present&nbsp;in&nbsp;scans|N|C|R|C|C|C|C|C|C|C
ASENT7|TAGS_FOR_RELATIVE_SETLOC_AND_BLANK_BANK_CARDS|ST|Line&nbsp;not&nbsp;present&nbsp;in&nbsp;scans|N|C|R|C|C|C|C|C|C|C
RODTRAP|TAGS_FOR_RELATIVE_SETLOC_AND_BLANK_BANK_CARDS|ST|Line&nbsp;not&nbsp;present&nbsp;in&nbsp;scans|N|C|R|C|C|C|C|C|C|C
V96ONBIT|ERASABLE_ASSIGNMENTS|ST|Definition&nbsp;not&nbsp;present&nbsp;in&nbsp;transcription|N|N|N|A|A|A|A|A|C|C
RUPTSTOR|ERASABLE_ASSIGNMENTS|ST|Definition&nbsp;not&nbsp;present&nbsp;in&nbsp;scans|N|N|N|N|N|N|N|N|R|R
SUPERRSM|WAITLIST|ST|Symbol&nbsp;not&nbsp;present&nbsp;in&nbsp;transcription|N|N|N|N|N|N|N|N|A|A
COUNT\*&nbsp;$$/P37|P37,P70|CT|Lines&nbsp;present&nbsp;in&nbsp;scans,&nbsp;missing&nbsp;from&nbsp;transcription|N|N|N|N|N|N|N|N|C|L1082

**Comment by smithery1** [comment]
### CM Page Numbering Changes

Action|S55|C237|C249|C044\*|C045\*|M45R2\*|C051\*|C055|A071\*|A072
|-|-|-|-|-|-|-|-|-|-|-
Capitalize&nbsp;Page&nbsp;for&nbsp;242,244,348|X|||||||||
Add&nbsp;empty&nbsp;page&nbsp;882||||X|X|X|X|X||
Add&nbsp;empty&nbsp;page&nbsp;191|||||||||X|X

**Comment by smithery1** [comment]
### CM Other Changes

Other|Corrected|Basis|Location|S55|C237|C249|C044\*|C045\*|M45R2\*|C051\*|C055|A071\*|A072
|-|-|-|-|-|-|-|-|-|-|-|-|-|-
+8|+8D|A|ERASABLE_ASSIGNMENTS|||L1154|C|C|C|C|C|C|C
R53FLBIT=BIT8|R53FLBIT=BIT6|ET|ERASABLE_ASSIGNMENTS|||L444|L424|L419|L422|L424|L455|C|C
FCADRMM1|Move&nbsp;to&nbsp;separate&nbsp;EQUALS&nbsp;line&nbsp;to&nbsp;match&nbsp;scans|ET|FRESH_START_AND_RESTART|||L1369|C|C|C|C|C|C|
KTETA1|Fix&nbsp;single&nbsp;\*&nbsp;use|A|P61&#8209;P67|||||||||L1065|L1065

**Comment by smithery1** [comment]
Merged to main so closing this out.
